### PR TITLE
Code folding fixes

### PR DESCRIPTION
**Pull request by abose**
**Do not merge yet** - we should land PR #10859(Code folding UI & code style tweaks) first.

Fixes Unit test failures after code folding integration and issues #10857 , #10856 

point 6 in issue #10857 is still fix in progress.

### CODE REVIEW

**Comment by abose** [comment]
The only change is in /foldhelpers/foldgutter.js, All other changes are part of PR #10859

**Comment by thehogfather** [comment]
This looks a good fix for the phantom fold marks issue.

**Comment by abose** [comment]
@peterflynn waiting for PR #10859 to be merged.

**Comment by peterflynn** [comment]
@abose I think there might be something funny with the git history in this PR -- if I view a [comparison against my branch](https://github.com/adobe/brackets/compare/pflynn/codefolding-tweaks...CodeFoldingFixes), it appears to also contain a bunch of ALF changes.  Please check to make sure the history is in an ok state before anyone merges this.

Also, in the future please use `abose/` as a prefix on your branch names -- since all committers are pushing branches to this repo, having informal "namespaces" is important to help keep branches organized.

File: src/extensions/default/CodeFolding/foldhelpers/foldgutter.js [hunk 1]
```diff
@@ -214,7 +211,7 @@ define(function (require, exports, module) {
             var lineChanges = changeObj.text.length - changeObj.removed.length;
             //update the lineFolds cache
             updateFoldsCache(cm, changeObj.from.line, lineChanges);
-            if (lineChanges !== 0) {
+            if (lineChanges !== 0 && changeObj.from.line + lineChanges >= 0) {
```
Comment (peterflynn) [inline src/extensions/default/CodeFolding/foldhelpers/foldgutter.js]: I'm not sure I understand this part of the change. The only time `changeObj.from.line + lineChanges < 0` could happen is when a bunch of lines were deleted at the start of the file.  Do we actually want to skip running `updateFoldInfo()` in that case?

I'm wondering if what we really want is something more like `Math.max(0, changeObj.from.line + lineChanges)` on the line below this instead... and similar for the 2nd arg too.

**Comment by abose** [comment]
changed, `Math.max(0, changeObj.from.line + lineChanges)` and calling `updateFoldInfo()` sounds more logical.

**Comment by abose** [comment]
point 6 in issue #10857 is Also fixed with the change.

**Comment by abose** [comment]
Thanks @thehogfather for verifying.
Merging the fix.

**Comment by peterflynn** [comment]
Ah, ok so @thehogfather got a chance to look at the additional changes here that occurred after my review?  There's a whole bunch of new code here that was added (`syncDocToFoldsCache()` and a bunch of other changes in commits 45c7a99b, d69ade37, & 684a507e)...  Normally that would require a re-review by someone before merging.

**Comment by thehogfather** [comment]
I'll get some more time to test and review more rigorously this evening and update the thread.

**Comment by abose** [comment]
@thehogfather Could you help with fixing some of these issues now that we are nearing the release date (probably next week), It would help very much if you could!

**Comment by thehogfather** [comment]
Yes I can. I'll have a look this evening.

**Comment by abose** [comment]
Thanks :)

**Pull request merged**